src/meili/config.cc

```diff
@@ -0,0 +1,113 @@
+#include "meili/config.h"
+#include "midgard/logging.h"
+
+#define CHECK_POSITIVE_PARAM(value, name)                                                            \
+  {                                                                                                  \
+    if (value <= 0.f)                                                                                \
+      throw std::invalid_argument(std::string("Expect '") + name +                                   \
+                                  "' to be positive (got: " + std::to_string(value) + ")");          \
+  }
+
+#define CHECK_NONNEGATIVE_PARAM(value, name)                                                         \
+  {                                                                                                  \
+    if (value < 0.f)                                                                                 \
+      throw std::invalid_argument(std::string("Expect '") + name +                                   \
+                                  "' to be nonnegative (got: " + std::to_string(value) + ")");       \
+  }
+
+namespace {
+template <typename T>
+inline void
+ReadParamOptional(T& value, const boost::property_tree::ptree& ptree, std::string const& name) {
+  if (auto item = ptree.get_optional<T>(name))
+    value = *item;
+}
+} // namespace
+
+namespace valhalla {
+namespace meili {
+
+Config::Config(const boost::property_tree::ptree& params) {
+  Read(params);
+}
+
+void Config::Read(const boost::property_tree::ptree& params) {
+  candidate_search.Read(params);
+  transition.Read(params);
+  emission.Read(params);
+  routing.Read(params);
+  customizable.Read(params);
+}
+
+void Config::CandidateSearch::Read(const boost::property_tree::ptree& params) {
+  ReadParamOptional(search_radius, params, "default.search_radius");
```

Comment:
I'm not sure about this: should we treat parameters from json as optional or required ?

Reply:
I _think_ this it should be optional, with a sensible default configuration. My understanding though is that this is different from the current behavior, but should be fine because we are relaxing (vs strengthening) an API constraint (i.e., what was previously optional is now required).